two phases flow codes/Dambreak_LevelSet_script.py

```diff
@@ -0,0 +1,239 @@
+
+V = FunctionSpace(mesh, 'P', 1)
+
+
+## Define initial conditions
+epsilon=0.5*pow(scale/res,0.9)#related to the size of element
+interface_x=0.3*scale
+interface_y=0.5*scale
+initial_phi_c = CharacterLevelSetFunction(param1=interface_x, param2=interface_y, param3=epsilon, element=V.ufl_element(),degree=1)#initial level set function
+initial_v = Expression(('0.', '0.'), degree=1)
+initial_p = Expression('(x[0] < x0 && x[1] < y0) ? (y0-x[1])*1.0 : 0.0',x0=interface_x,y0=interface_y, degree=1) #initial the pressure field of water 
+
+#%%
+## initialization
+
+# project(initial_phi_c, ScalarSpace1)
+# project(initial_v, VectorSpace)
+# project(initial_p, ScalarSpace)
+phi_c.assign(project(initial_phi_c, ScalarSpace1))
+phi_c0.assign(project(initial_phi_c, ScalarSpace1))
+u_c0.assign(project(initial_v, VectorSpace))
+p0.assign(project(initial_p*initial_phi_c, ScalarSpace)) 
+#mutiply the inital level-set value can make the initial calculation step more stable, but it has no effect on the result
+# can also use p0.assign(project(initial_p, ScalarSpace)) 
+
+
+# set parameters
+Re=2000.
+Fr=0.7
+rho0=1.
+nu0=1.
+rho1=0.001
+nu1=1.
+# write file
+file_pp = File(foldername+'/p_Re%d' %Re + '_Fr%.3s.pvd'%Fr)
+file_vv = File(foldername+'/v_Re%d' %Re + '_Fr%.3s.pvd'%Fr)
+file_c = File(foldername+'/c_Re%d' %Re + '_Fr%.3s.pvd'%Fr)
+file_c << (phi_c0,t)
+file_pp << (p0,t)
+
+file_p = XDMFFile(foldername+'/p_Re%d' %Re + '_Fr%.3s.xdmf'%Fr)
+file_v = XDMFFile(foldername+'/v_Re%d' %Re + '_Fr%.3s.xdmf'%Fr)
+file_phi = XDMFFile(foldername+'/phi_Re%d' %Re + '_Fr%.3s.xdmf'%Fr)
+#file_p.write_checkpoint(p, "p", t, append=False)
+#file_v.write_checkpoint(u_c, "v", t, append=False)
+#file_phi.write_checkpoint(phi_c, "phi", t, append=False)
+
+
+g_=-1.
+rou= phi_c*rho0+(1-phi_c)*rho1
+mu= phi_c*nu0+(1-phi_c)*nu1
+
+ndim = mesh.geometry().dim()
+
+# viscous operator
+def a(phi,chi, psi): return Constant(0.5) * inner(grad(phi) + grad(chi).T, grad(psi) + grad(psi).T)
+# divergence operator
+def b(phi, psi): return inner(div(phi), psi)
+# non-linear convection operator
+def c(phi, chi, psi): return dot(dot(grad(chi), phi), psi)
+# Crank-Nicholson schema average
+def d(phi, psi): return Constant(0.5) * (phi + psi)
+
+#n_phi=grad(phi_c)/pow(dot( grad(phi_c), grad(phi_c) ), 0.5)
+g = Expression(('0.','g'),degree=1,g=g_)
+
+# interface normal n = ▽ϕ/|ϕ|
+F_n_phi = (dot(grad(phi_c)/pow(dot( grad(phi_c), grad(phi_c) ), 0.5), del_n_phi) - dot(n_phi,del_n_phi)) * dx
+
+## Projection method
+F_momentum = (rou*dot((u_t-u_c0),v)/dt + rou*c(u_t, u_t, v) + (1. / Re) * mu*a(u_t,u_t, v) -  b(v, p0) - (1./(Fr*Fr))*rou*(dot(g,v))) * dx
+#F_momentum = (rou*dot((u_t-u_c0),v)/dt - dot(u_t, div(rou*outer(v,u_t)) ) + (1. / Re) * mu*a(u_t,u_t, v) -  b(v, p0) - (1./(Fr*Fr))*rou*(dot(g,v))) * dx
+# pressure correction
+F_pressure = (inner(grad(p-p0),grad(q))/rou + (1/dt)*div(u_t)*q) * dx
+#velocity correction
+F_velocity = ((1/dt)*inner(u_c-u_t,v) + inner(grad(p-p0),v)/rou) * dx
```

Comment:
We should write a solver for the ICPC scheme anyway. This is already on my ToDo list.

Reply:
What is ICPC scheme? I'm sorry I don't know much about this, are there any references?

Reply:
Sorry this was a typo. It should IPCS which means incremental pressure correction scheme. This is the scheme you are using to solve the Navier-Stokes problem. Are you familiar with incremental pressure correction schemes?

Reply:
I am not familiar with this, I can learn it.